pypeit/images/rawimage.py

```diff
@@ -19,399 +19,701 @@
+                construct the slit illumination profile, if requested via the
+                ``use_illumflat`` parameter in :attr:`par`; see
+                :func:`~pypeit.flatfield.FlatImages.fit2illumflat`.
+            debug (:obj:`bool`, optional):
+                Run in debug mode.
+            dark (:class:`~pypeit.images.pypeitimage.PypeItImage`):
+                Dark image
 
         Returns:
-            :class:`pypeit.images.pypeitimage.PypeItImage`:
+            :class:`~pypeit.images.pypeitimage.PypeItImage`:
 
         """
+        # TODO: Reset steps to all be false at the beginning of the function?
+        # If we're careful with book-keeping of the attributes (i.e., resetting
+        # image to rawimage, etc), calling process multiple times could be a way
+        # for developers to *re*process an image multiple times from scratch
+        # with different parameters to test changes.
+
+        # Set input to attributes
         self.par = par
-        self._bpm = bpm
-        
-        # Get started
-        # Standard order
-        #   -- May need to allow for other order some day..
-        if par['use_pattern']:  # Note, this step *must* be done before use_overscan
+        if bpm is not None:
+            self._bpm = bpm
+
+        # Check the input
+        if self.par['use_biasimage'] and bias is None:
+            msgs.error('No bias available for bias subtraction!')
+        if self.par['use_darkimage'] and dark is None:
+            msgs.error('No dark available for dark subtraction!')
+        if self.par['spat_flexure_correct'] and slits is None:
+            msgs.error('Spatial flexure correction requested but no slits provided.')
+        if self.use_flat and flatimages is None:
+            msgs.error('Flat-field corrections requested but no flat-field images generated '
+                       'or provided.  Make sure you have flat-field images in your PypeIt file!')
+
+        # Apply additive corrections.  The order matters and is fixed.
+        #   - Subtract any fixed pattern defined for the instrument.  NOTE: This
+        #     step *must* be done before use_overscan
+        if self.par['use_pattern']:
             self.subtract_pattern()
-        if par['use_overscan']:
+
+        #   - Estimate the readnoise using the overscan regions.  NOTE: I'm
+        #     assuming this needs to be done *after* the sine pattern has been
+        #     subtracted.
+        if not np.all(self.ronoise > 0):
+            self.estimate_readnoise()
+
+        #   - Get detector variance image.  Starts with the shape and
+        #     orientation of the raw image, and the units are in ADU to match
+        #     the current units of the image being processed.  Note that the
+        #     readnoise variance image is kept separate because it is used again
+        #     when adding the shot noise.
+        self.rn2img = self.build_rn2img(units='ADU')
```

Comment:
We should never ever be estimating noise in ADU. I think this is the wrong approach entirely and will only lead to noise and confusion. It also goes against the standard way in which read noise is quoted, which is always in electrons.